`azurerm_key_vault_secret` - allow Key Vault Reader to make plan of secret resources with value_wo

abicky opened the pull request:
<!--  All Submissions -->


## Community Note
<!-- Please leave the community note as is. -->
* Please vote on this PR by adding a :thumbsup: [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original PR to help the community and maintainers prioritize for review
* Please do not leave comments along the lines of "+1", "me too" or "any updates", they generate extra noise for PR followers and do not help prioritize for review


## Description

<!-- Please include a description below with the reason for the PR, what it is doing, what it is trying to accomplish, and anything relevant for a reviewer to know. 

If this is a breaking change for users please detail how it cannot be avoided and why it should be made in a minor version of the provider -->

Previously, `terraform plan` required permission to read secret values, even for resources using `value_wo`, which do not need access to the actual secret value.
This PR relaxes that requirement, allowing users with the Key Vault Reader role, which does not have permission to read secret values, to run `terraform plan` with minimal permissions.

## PR Checklist

- [x] I have followed the guidelines in our [Contributing Documentation](../blob/main/contributing/README.md).
- [x] I have checked to ensure there aren't other open [Pull Requests](../pulls) for the same update/change.
- [x] I have checked if my changes close any open issues. If so please include appropriate [closing keywords](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword) below.
- [x] I have updated/added Documentation as required written in a helpful and kind way to assist users that may be unfamiliar with the resource / data source.
- [x] I have used a meaningful PR title to help maintainers and other users understand this change and help prevent duplicate work. 
For example: “`resource_name_here` - description of change e.g. adding property `new_property_name_here`”


<!-- You can erase any parts of this template below this point that are not applicable to your Pull Request. -->


## Changes to existing Resource / Data Source

- [x] I have added an explanation of what my changes do and why I'd like you to include them (This may be covered by linking to an issue above, but may benefit from additional explanation).
- [x] I have written new tests for my resource or datasource changes & updated any relevant documentation.
- [x] I have successfully run tests with my changes locally. If not, please provide details on testing challenges that prevented you running the tests.
- [ ] (For changes that include a **state migration only**). I have manually tested the migration path between relevant versions of the provider.


## Testing 

- [x] My submission includes Test coverage as described in the [Contribution Guide](../blob/main/contributing/topics/guide-new-resource.md) and the tests pass. (if this is not possible for any reason, please include details of why you did or could not add test coverage)

<!-- Please include testing logs or evidence here or an explanation on why no testing evidence can be provided. 

For state migrations please test the changes locally and provide details here, such as the versions involved in testing the migration path. For further details on testing state migration changes please see our guide on [state migrations](https://github.com/hashicorp/terraform-provider-azurerm/blob/main/contributing/topics/guide-state-migrations.md#testing) in the contributor documentation. -->

```console
$ make acctests SERVICE="keyvault" TESTARGS='-skip=disappearsWhenParentKeyVaultDeleted -run=TestAccKeyVaultSecret'
TF_ACC=1 go test -v ./internal/services/keyvault -skip=disappearsWhenParentKeyVaultDeleted -run=TestAccKeyVaultSecret -timeout 180m -ldflags="-X=github.com/hashicorp/terraform-provider-azurerm/version.ProviderVersion=acc"
=== RUN   TestAccKeyVaultSecret_basic
=== PAUSE TestAccKeyVaultSecret_basic
=== RUN   TestAccKeyVaultSecret_writeOnlyValue
=== PAUSE TestAccKeyVaultSecret_writeOnlyValue
=== RUN   TestAccKeyVaultSecret_updateToWriteOnlyValue
=== PAUSE TestAccKeyVaultSecret_updateToWriteOnlyValue
=== RUN   TestAccKeyVaultSecret_writeOnlyValueWithoutReadPermission
=== PAUSE TestAccKeyVaultSecret_writeOnlyValueWithoutReadPermission
=== RUN   TestAccKeyVaultSecret_requiresImport
=== PAUSE TestAccKeyVaultSecret_requiresImport
=== RUN   TestAccKeyVaultSecret_disappears
=== PAUSE TestAccKeyVaultSecret_disappears
=== RUN   TestAccKeyVaultSecret_disappearsWriteOnlyValue
=== PAUSE TestAccKeyVaultSecret_disappearsWriteOnlyValue
=== RUN   TestAccKeyVaultSecret_complete
=== PAUSE TestAccKeyVaultSecret_complete
=== RUN   TestAccKeyVaultSecret_update
=== PAUSE TestAccKeyVaultSecret_update
=== RUN   TestAccKeyVaultSecret_updatingValueChangedExternally
=== PAUSE TestAccKeyVaultSecret_updatingValueChangedExternally
=== RUN   TestAccKeyVaultSecret_recovery
=== PAUSE TestAccKeyVaultSecret_recovery
=== RUN   TestAccKeyVaultSecret_withExternalAccessPolicy
=== PAUSE TestAccKeyVaultSecret_withExternalAccessPolicy
=== RUN   TestAccKeyVaultSecret_purge
=== PAUSE TestAccKeyVaultSecret_purge
=== CONT  TestAccKeyVaultSecret_basic
=== CONT  TestAccKeyVaultSecret_complete
=== CONT  TestAccKeyVaultSecret_recovery
=== CONT  TestAccKeyVaultSecret_updatingValueChangedExternally
=== CONT  TestAccKeyVaultSecret_requiresImport
=== CONT  TestAccKeyVaultSecret_disappearsWriteOnlyValue
=== CONT  TestAccKeyVaultSecret_disappears
=== CONT  TestAccKeyVaultSecret_purge
--- PASS: TestAccKeyVaultSecret_disappears (250.44s)
=== CONT  TestAccKeyVaultSecret_updateToWriteOnlyValue
--- PASS: TestAccKeyVaultSecret_disappearsWriteOnlyValue (256.21s)
=== CONT  TestAccKeyVaultSecret_writeOnlyValueWithoutReadPermission
--- PASS: TestAccKeyVaultSecret_purge (275.38s)
=== CONT  TestAccKeyVaultSecret_withExternalAccessPolicy
--- PASS: TestAccKeyVaultSecret_requiresImport (282.22s)
=== CONT  TestAccKeyVaultSecret_writeOnlyValue
--- PASS: TestAccKeyVaultSecret_basic (284.42s)
=== CONT  TestAccKeyVaultSecret_update
--- PASS: TestAccKeyVaultSecret_complete (286.98s)
--- PASS: TestAccKeyVaultSecret_updatingValueChangedExternally (305.99s)
--- PASS: TestAccKeyVaultSecret_recovery (520.78s)
--- PASS: TestAccKeyVaultSecret_writeOnlyValueWithoutReadPermission (267.91s)
--- PASS: TestAccKeyVaultSecret_updateToWriteOnlyValue (278.06s)
--- PASS: TestAccKeyVaultSecret_writeOnlyValue (267.57s)
--- PASS: TestAccKeyVaultSecret_update (279.92s)
--- PASS: TestAccKeyVaultSecret_withExternalAccessPolicy (313.85s)
PASS
ok      github.com/hashicorp/terraform-provider-azurerm/internal/services/keyvault      591.703s
```


## Change Log

Below please provide what should go into the changelog (if anything) conforming to the [Changelog Format documented here](../blob/main/contributing/topics/maintainer-changelog.md).

<!-- Replace the changelog example below with your entry. One resource per line. -->

* `azurerm_key_vault_secret` - allow Key Vault Reader to make plan of secret resources with value_wo


<!-- What type of PR is this? -->
This is a (please select all that apply):

- [ ] Bug Fix
- [ ] New Feature (ie adding a service, resource, or data source)
- [x] Enhancement
- [ ] Breaking Change


## Related Issue(s)
Fixes https://github.com/hashicorp/terraform-provider-azurerm/issues/30874


> [!NOTE] 
> If this PR changes meaningfully during the course of review please update the title and description as required.

abicky committed:
`azurerm_key_vault_secret` - allow Key Vault Reader to make plan

Previously, `terraform plan` required permission to read secret values,
even for resources using `value_wo`, which do not need access to the
actual secret value.
This commit relaxes that requirement, allowing users with the Key Vault
Reader role to run `terraform plan` with minimal permissions.

gerrytan commented on internal/services/keyvault/key_vault_secret_resource.go:
```diff
@@ -496,3 +519,59 @@ func (d deleteAndPurgeSecret) NestedItemHasBeenPurged(ctx context.Context) (auto
+	}
+
+	return secret, nil
+}
```
I wonder if it's possible to use https://github.com/hashicorp/go-azure-sdk/tree/main/data-plane/keyvault instead here. If you merge latest main (and resolve the conflict), this module is now available.

abicky replied:
Thanks for the suggestion! I don't think we can use the data-plane Key Vault client here.

As far as I can tell, there isn't a [Key Vault REST API](https://learn.microsoft.com/en-us/rest/api/keyvault/) that returns a secret without its value. The closest endpoint is the [Get Secret Versions API](https://learn.microsoft.com/en-us/rest/api/keyvault/secrets/get-secret-versions?view=rest-keyvault-secrets-2025-07-01), which returns version metadata without the secret value. However, we'd need to enumerate all versions to determine the latest one, similar to the implementation here:
https://github.com/abicky/terraform-provider-azurekv/blob/v0.1.2/internal/provider/client.go#L57-L89

Given that, I think using the Resources client is the best approach for this case.

abicky replied:
By the way, I've rebased the branch and added a refactoring, along with another commit introducing the `set_value_wo_version_on_import` flag, which is required to import secrets without requiring permission to read the secret value.

gerrytan reviewed (changes requested):
Thank you for the contrib @abicky. I think this is a good improvement. I left a question / suggestion, can you please take a look. I'll help running the acctest and the workflow if we can proceed with this.

abicky committed:
`azurerm_key_vault_secret` - add set_value_wo_version_on_import flag

This commit adds the set_value_wo_version_on_import flag to allow
importing azurerm_key_vault_secret resources with value_wo_version
enabled without requiring permission to read the secret value.

gerrytan commented on internal/services/keyvault/key_vault_secret_resource.go:
```diff
@@ -501,3 +497,65 @@ func (d deleteAndPurgeSecret) NestedItemHasBeenPurged(ctx context.Context) (auto
+	prop, ok := (*resp.Model.Properties).(map[string]any)
+	if !ok {
+		return secret, fmt.Errorf("getting Key Vault Secret %q: unexpected property type %T", resourceID, *resp.Model.Properties)
+	}
+
+	if v, ok := prop["secretUriWithVersion"].(string); ok {
+		secret.ID = &v
+	}
+	if v, ok := prop["contentType"].(string); ok {
+		secret.ContentType = &v
+	}
+	if attrs, ok := prop["attributes"].(map[string]any); ok {
+		secret.Attributes = &kv.SecretAttributes{}
+		if v, ok := attrs["enabled"].(bool); ok {
+			secret.Attributes.Enabled = &v
+		}
+		if v, ok := attrs["created"].(float64); ok {
+			t := date.NewUnixTimeFromSeconds(v)
+			secret.Attributes.Created = &t
+		}
+		if v, ok := attrs["updated"].(float64); ok {
+			t := date.NewUnixTimeFromSeconds(v)
+			secret.Attributes.Updated = &t
+		}
+		if v, ok := attrs["exp"].(float64); ok {
+			t := date.NewUnixTimeFromSeconds(v)
+			secret.Attributes.Expires = &t
+		}
+		if v, ok := attrs["nbf"].(float64); ok {
+			t := date.NewUnixTimeFromSeconds(v)
+			secret.Attributes.NotBefore = &t
+		}
+	}
+
+	if resp.Model.Tags != nil {
+		secret.Tags = tags.FromTypedObject(*resp.Model.Tags)
+	}
+
+	return secret, nil
+}
```
Thx @abicky , I think this PR is a good improvement. I'm just a bit concerned with all the heavy lifting the provider has to do here.

I think I'll check with @jackofallops first if there's a way this extra work can be done inside go-azure-sdk/data-plane rather than here.
